File: src/main/java/cn/solarmoon/idyllic_food_diary/compat/jade/JadeUtil.java
package cn.solarmoon.idyllic_food_diary.compat.jade;

import cn.solarmoon.idyllic_food_diary.IdyllicFoodDiary;
import cn.solarmoon.idyllic_food_diary.element.matter.cookware.steamer.SteamerBlockEntity;
import cn.solarmoon.idyllic_food_diary.feature.basic_feature.IPlateable;
import cn.solarmoon.idyllic_food_diary.feature.generic_recipe.evaporation.IEvaporationRecipe;
import cn.solarmoon.idyllic_food_diary.feature.generic_recipe.ingredient_handling.IIngredientHandlingRecipe;
import cn.solarmoon.idyllic_food_diary.feature.generic_recipe.stir_fry.IStirFryRecipe;
import net.minecraft.ChatFormatting;
import net.minecraft.network.chat.Component;
import net.minecraft.util.StringUtil;
import net.minecraftforge.items.IItemHandler;
import snownee.jade.api.ITooltip;
import snownee.jade.api.ui.BoxStyle;
import snownee.jade.api.ui.IElement;
import snownee.jade.api.ui.IElementHelper;
import snownee.jade.impl.ui.ProgressStyle;

public class JadeUtil {

    public static void addSteamingBaseTip(ITooltip iTooltip, IEvaporationRecipe eva) {
        IElementHelper ehp = iTooltip.getElementHelper();
        if (eva.isEvaporating()) {
            iTooltip.add(ehp.text(IdyllicFoodDiary.TRANSLATOR.set("jade", "evaporating", eva.getEvaporatingAmount())));
        }
    }

    public static void addSteamingTip(ITooltip iTooltip, SteamerBlockEntity steamer) {
        IElementHelper ehp = iTooltip.getElementHelper();

        int stackAmount = steamer.getStack();
        for (int i = 0; i < stackAmount; i++) {
            int r = i + 1;
            int index = i == 0 ? iTooltip.size() : iTooltip.size() - i;
            iTooltip.add(index, IdyllicFoodDiary.TRANSLATOR.set("jade", "steamer.layer_" + r));
            if (i < steamer.getInvList().size()) {
                steamer.getInvList().get(i).getStacks().forEach(stack -> iTooltip.append(index, ehp.item(stack, 0.5f)));
            }
        }

        iTooltip.add(IdyllicFoodDiary.TRANSLATOR.set("jade", "steamer.working"));
        iTooltip.append(Component
                .literal(steamer.canWork() ? "✔" : "✖")
                .withStyle(steamer.canWork() ? ChatFormatting.GREEN : ChatFormatting.RED)
        );
    }

    public static void addIngredientHandlingResult(ITooltip iTooltip, IIngredientHandlingRecipe ih) {
        IElementHelper ehp = iTooltip.getElementHelper();
        iTooltip.add(ehp.text(
                IdyllicFoodDiary.TRANSLATOR.set("jade", "ingredient_handling")
                        .copy().append(
                                ih.hasOutput() ?
                                        Component.translatable(ih.findHandleRecipe().get().result().getDescriptionId())
                                        : IdyllicFoodDiary.TRANSLATOR.set("jade", "ingredient_handling.empty")
                        )
        ));
        if (ih.hasOutput()) iTooltip.append(ehp.item(ih.findHandleRecipe().get().result(), 0.5f));
    }

    public static void addStirFryStage(ITooltip iTooltip, IStirFryRecipe stir) {
        IElementHelper ehp = iTooltip.getElementHelper();
        iTooltip.add(ehp.text(IdyllicFoodDiary.TRANSLATOR.set("jade", "stir_fry.stage",
                        stir.isStirFrying() ? stir.getPresentStage() + 1 : 0
                )
        ));
    }

    public static void addPlatingResult(ITooltip iTooltip, IPlateable pr) {
        IElementHelper ehp = iTooltip.getElementHelper();
        iTooltip.add(ehp.text(IdyllicFoodDiary.TRANSLATOR.set("jade", "plating").copy().append(
                        pr.getResult().isEmpty() ?
                                IdyllicFoodDiary.TRANSLATOR.set("jade", "plating.empty") :
                                Component.translatable(pr.getResult().getDescriptionId())
                )
        ));
        if (!pr.getResult().isEmpty()) iTooltip.append(ehp.item(pr.getResult(), 0.5f));
    }

    public static void addByTime(int timeNow, int recipeTime, ITooltip iTooltip) {
        float scale = (float) timeNow / recipeTime;
        if (timeNow != 0) {
            IElementHelper ehp = iTooltip.getElementHelper();
            IElement progress = ehp.progress(
                    scale,
                    Component.literal(StringUtil.formatTickDuration(timeNow) + "/" + StringUtil.formatTickDuration(recipeTime)).withStyle(ChatFormatting.WHITE),
                    new ProgressStyle(),
                    BoxStyle.DEFAULT,
                    true
            );
            iTooltip.add(progress);
        }
    }

    public static void addByTimeArray(int[] times, int[] recipeTimes, IItemHandler inv, ITooltip iTooltip) {
        int n = 0;
        for (int i = 0; i < times.length; i ++) {
            if (times[i] != 0) {
                int time = times[i];
                int needTime = recipeTimes[i];
                IElementHelper ehp = iTooltip.getElementHelper();
                if (n % 2 == 0) {
                    iTooltip.add(ehp.smallItem(inv.getStackInSlot(i)));
                } else {
                    iTooltip.append(iTooltip.size() - 1, ehp.smallItem(inv.getStackInSlot(i)));
                }
                n++;
                iTooltip.append(Component.literal(StringUtil.formatTickDuration(time) + "/" + StringUtil.formatTickDuration(needTime)).withStyle(ChatFormatting.WHITE));
            }
        }
    }

}

File: src/main/kotlin/cn/solarmoon/idyllic_food_diary/compat/jade/JadeUtil.kt
package cn.solarmoon.idyllic_food_diary.compat.jade

import cn.solarmoon.idyllic_food_diary.IdyllicFoodDiary
import cn.solarmoon.idyllic_food_diary.element.matter.cookware.steamer.SteamerBlockEntity
import cn.solarmoon.idyllic_food_diary.element.recipe.EvaporationRecipe
import cn.solarmoon.idyllic_food_diary.element.recipe.StirFryRecipe
import cn.solarmoon.idyllic_food_diary.element.recipe.assistant.IPlateable
import cn.solarmoon.spark_core.api.cap.item.ItemStackHandlerHelper
import net.minecraft.ChatFormatting
import net.minecraft.core.Direction
import net.minecraft.network.chat.Component
import net.minecraft.util.StringUtil
import net.minecraft.world.phys.BlockHitResult
import snownee.jade.api.ITooltip
import snownee.jade.api.ui.BoxStyle.GradientBorder
import snownee.jade.impl.ui.ItemStackElement
import snownee.jade.impl.ui.ProgressElement
import snownee.jade.impl.ui.SimpleProgressStyle
import snownee.jade.impl.ui.TextElement

object JadeUtil {

    @JvmStatic
    fun addByTime(timeNow: Int, recipeTime: Int, iTooltip: ITooltip) {
        val scale = timeNow / recipeTime.toFloat()
        if (timeNow != 0) {
            val progress = ProgressElement(
                scale,
                Component.literal(StringUtil.formatTickDuration(timeNow, 20f) + "/" + StringUtil.formatTickDuration(recipeTime, 20f)).withStyle(ChatFormatting.WHITE),
                SimpleProgressStyle(),
                GradientBorder.DEFAULT_VIEW_GROUP,
                true
            )
            iTooltip.add(progress)
        }
    }

    @JvmStatic
    fun addStirFryStage(iTooltip: ITooltip, stir: StirFryRecipe.Processor) {
        val text = TextElement(IdyllicFoodDiary.TRANSLATOR.set("jade", "stir_fry.stage",
            if (stir.isWorking()) stir.presentStageIndex + 1 else 0))
        iTooltip.add(text)
    }

    @JvmStatic
    fun addPlatingResult(iTooltip: ITooltip, pr: IPlateable) {
        val text = TextElement(IdyllicFoodDiary.TRANSLATOR.set("jade", "plating").copy().append(
            if (pr.result.isEmpty) IdyllicFoodDiary.TRANSLATOR.set("jade", "plating.empty")
            else Component.translatable(pr.result.descriptionId)
        ))
        iTooltip.add(text)
        val item = ItemStackElement.of(pr.result, 0.5f)
        if (!pr.result.isEmpty) iTooltip.append(item)
    }

    @JvmStatic
    fun addSteamingTip(iTooltip: ITooltip, steamer: SteamerBlockEntity, hit: BlockHitResult) {
        val dy = hit.getLocation().y - steamer.blockPos.y
        var layer = (dy * 16 / steamer.maxLayer).toInt()
        if (hit.direction == Direction.UP) layer--
        val r = layer + 1
        if (layer < steamer.presentLayer) {
            iTooltip.add(IdyllicFoodDiary.TRANSLATOR.set("jade", "steamer.layer_$r"))
            ItemStackHandlerHelper.getStacks(steamer.inventories[layer]).forEach { stack -> iTooltip.append(ItemStackElement.of(stack, 0.5f)) }
        }

        iTooltip.add(IdyllicFoodDiary.TRANSLATOR.set("jade", "steamer.working"))
        iTooltip.append(Component
            .literal(if (steamer.steaming.canWork()) "✔" else "✖").withStyle(if (steamer.steaming.canWork()) ChatFormatting.GREEN else ChatFormatting.RED)
        )
    }

    @JvmStatic
    fun addEvaporationTip(iTooltip: ITooltip, eva: EvaporationRecipe.Processor) {
        if (eva.isWorking()) {
            iTooltip.add(TextElement(IdyllicFoodDiary.TRANSLATOR.set("jade", "evaporating", eva.getEvaporatingAmount())))
        }
    }

}